002-es-modules.md

```diff
@@ -0,0 +1,175 @@
+      prepopulated.
+
+## 3. Semantics
+
+### 3.1. Determining if source is an ES Module
+
+A new file type will be used for Node v2.0 Modules, `.mjs`. Files with this extension
+will be treated using the loading semantics in this document, but retain a way
+to load existing extensions of `.json`, `.node`, and `.js`. The file type for
+JavaScript with IANA as an official file type needs to have this extensions
+added. This comes from browsers being unable to implement a new MIME for the
+differing parse goals. In order to make this change, contact the IESG.
+
+The `.mjs` file extension will be searched for prior to any `.js` file in any
+algorithm that uses searching (`require`); e.g. once the Node
+resolution algorithm reaches file expansion, `path + '.mjs'` would be searched
+prior to `path + '.js'` when performing `require(path)`.
+
+# 3.1.1. Inter package loading using file extension breakage.
+
+There is knowledge of breakage for code that upgrades inner package
+dependencies such as `require('foo/bar.js')`. As `bar.js` may move to
+`bar.mjs`. Since `bar.js` is not the listed entry point this was considered
+acceptable.
+
+### 3.2. Importing Modules in NM2
+
+Node v2.0 Modules do not implement `import` and `export`.
+Instead, modules are "imported" as usual and as identically as possible as NCJS,
+via `require()`.
+
+### 3.2.1 Function Wrap Approach
+
+To achieve this, modules would be wrapped in a function like a regular script
+file, but evaluated as a Module under the Module Parse Goal. However, that
+wrapper _may_ be exported with an `export` statement under the hood. As such,
+it _may_ look like the following:
+
+```js
+export default function (exports, require, module, __filename, __dirname) {
+  // user code inserted here
+}
+```
+
+_If_ exporting under the hood is required, the resulting (hidden) promise
+would likely be resolved synchronously on `require()`, given we are in
+complete control and can guarantee its safety.
+
+### 3.4. NM2 Evaluation
+
+#### 3.4.1. Environment Variables
+
+NM2 will be bootstrapped with standard NCJS "magic" variables.
+
+| Variable | Exists | Value |
+| ---- | ---- | --- |
+| this | y | ? |
+| arguments | y | same as regular scripts |
+| require | y | same as regular scripts |
+| module | y | same as regular scripts |
+| exports | y | same as regular scripts |
+| __filename | y | same as regular scripts |
+| __dirname | y | same as regular scripts |
+
+Like normal scoping rules, if a variable does not exist in a scope, the outer scope is used to find the variable. Since NM2 are always strict, errors may be thrown upon trying to use variables that do not exist globally when using NM2.
+
+### 4. Caveats
+
+This approach saves existing operability and tooling functionality in exchange
+for a couple caveats, namely:
+
+1. `import` & `export` are unimplemented / unavailable to users
+2. Top-level scope is unavailable.
+  - As such, top-level `await` is unavailable. (Also for sync safety reasons.)
```

Comment:
I'm saying that exceptions do not bubble inside async functions just like they don't in Promises. Top-level await, however, would have that functionality, because it's not inside an async function.